src/components/shapes/draw.js

```diff
@@ -579,6 +600,249 @@ function setupDragElement(gd, shapePath, shapeOptions, index, shapeLayer, editHe
+                    else if(shapex1 < shapex0) xanchor = 'right';
+                    else xanchor = 'center';
+                }
+            } else {
+                xanchor = 'center';
+            }
+        }
+
+        // Special case for padding when angle is 'auto' for lines
+        // Padding should be treated as an orthogonal offset in this case
+        // Otherwise, padding is just a simple x and y offset
+        var paddingConstantsX = { left: 1, center: 0, right: -1 };
+        var paddingConstantsY = { bottom: -1, middle: 0, top: 1 };
+        if(textAngle === 'auto') {
+            // Set direction to apply padding (based on `yanchor` only)
+            var paddingDirection = paddingConstantsY[yanchor];
+            paddingX = -textPadding * Math.sin(textAngleRad) * paddingDirection;
+            paddingY = textPadding * Math.cos(textAngleRad) * paddingDirection;
+        } else {
+            // Set direction to apply padding (based on `xanchor` and `yanchor`)
+            var paddingDirectionX = paddingConstantsX[xanchor];
+            var paddingDirectionY = paddingConstantsY[yanchor];
+            paddingX = textPadding * paddingDirectionX;
+            paddingY = textPadding * paddingDirectionY;
+        }
+        textx = textx + paddingX;
+        texty = texty + paddingY;
+    } else {
+        // Text position for shapes that are not lines
+        // calc horizontal position
+        // Horizontal needs a little extra padding to look balanced
+        paddingX = textPadding + 3;
+        if(textPosition.indexOf('right') !== -1) {
+            textx = Math.max(shapex0, shapex1) - paddingX;
+            if(xanchor === 'auto') xanchor = 'right';
+        } else if(textPosition.indexOf('left') !== -1) {
+            textx = Math.min(shapex0, shapex1) + paddingX;
+            if(xanchor === 'auto') xanchor = 'left';
+        } else { // Default: center
+            textx = (shapex0 + shapex1) / 2;
+            if(xanchor === 'auto') xanchor = 'center';
+        }
+
+        // calc vertical position
+        if(textPosition.indexOf('top') !== -1) {
+            texty = Math.min(shapey0, shapey1);
+        } else if(textPosition.indexOf('bottom') !== -1) {
+            texty = Math.max(shapey0, shapey1);
+        } else {
+            texty = (shapey0 + shapey1) / 2;
+        }
+        // Apply padding
+        paddingY = textPadding;
+        if(yanchor === 'bottom') {
+            texty = texty - paddingY;
+        } else if(yanchor === 'top') {
+            texty = texty + paddingY;
+        }
+    }
+
+    // Shift vertical (& horizontal) position according to `yanchor`
+    var shiftFraction = FROM_TL[yanchor];
+    // Adjust so that text is anchored at top of first line rather than at baseline of first line
+    var baselineAdjust = shapeOptions.label.font.size;
+    var textHeight = textBB.height;
+    var xshift = (textHeight * shiftFraction - baselineAdjust) * Math.sin(textAngleRad);
```

Comment:
It would be nice to calculate these slow Math functions once in this function and then reuse the results where needed.
```js
var sinA = Math.sin(textAngleRad);
var cosA = Math.cos(textAngleRad);
...
```